crates/cxx-qt-gen/src/parser/parameter.rs

```diff
@@ -16,29 +20,67 @@ pub struct ParsedFunctionParameter {
+                    Ok(Some(parameter))
+                }
+                // Ignore self as a parameter
+                FnArg::Receiver(_) => Ok(None),
+            }
+        })
+        .filter_map(|result| result.map_or_else(|e| Some(Err(e)), |v| v.map(Ok)))
+        .collect::<Result<Vec<ParsedFunctionParameter>>>()
+    }
+
+    pub fn parse_all_ignoring_receiver(
+        signature: &Signature,
+    ) -> Result<Vec<ParsedFunctionParameter>> {
+        let mut iter = signature.inputs.iter();
+        // whilst we can ignore the receiver argument, make sure it actually exists
+        if iter.next().is_none() {
+            return Err(Error::new_spanned(signature, "Missing receiver argument!"));
+        }
+
+        Self::parse_remaining(iter)
+    }
+
+    /// This function parses the list of arguments
     pub fn parse_all_without_receiver(
         signature: &Signature,
     ) -> Result<Vec<ParsedFunctionParameter>> {
-        signature
-            .inputs
-            .iter()
-            .map(|input| {
-                match input {
-                    FnArg::Typed(type_pattern) => {
-                        let parameter = ParsedFunctionParameter::parse(type_pattern)?;
-
-                        // Ignore self as a parameter
-                        if parameter.ident == "self" {
-                            return Ok(None);
-                        }
-                        Ok(Some(parameter))
-                    }
-                    // Ignore self as a parameter
-                    FnArg::Receiver(_) => Ok(None),
+        let mut iter = signature.inputs.iter();
+
+        let missing_self_arg = "First argument must be a supported `self` receiver type!\nUse `&self` or `self: Pin<&mut Self>` instead.";
+        match iter.next() {
+            Some(FnArg::Typed(type_pattern)) => {
+                let parameter = ParsedFunctionParameter::parse(type_pattern)?;
+                if parameter.ident == "self" && types::is_pin_of_self(&parameter.ty) {
+                    Ok(())
+                } else {
+                    Err(Error::new_spanned(type_pattern, missing_self_arg))
                 }
-            })
-            .filter_map(|result| result.map_or_else(|e| Some(Err(e)), |v| v.map(Ok)))
-            .collect::<Result<Vec<ParsedFunctionParameter>>>()
+            }
+            Some(FnArg::Receiver(Receiver {
+                reference: Some(_), // `self` needs to be by-ref, by-value is not supported.
+                mutability: None,   // Mutable `&mut self` references are not supported.
+                                    // Use `Pin<&mut  Self>` instead.
+                ..
+            })) => Ok(()), // Okay, found a &self reference
+            Some(arg) => Err(Error::new_spanned(arg, missing_self_arg)),
+            None => Err(Error::new_spanned(signature, "Missing 'self' receiver!")),
+        }?;
+
+        Self::parse_remaining(iter)
     }
 
     pub fn parse(type_pattern: &PatType) -> Result<Self> {
```

Comment:
do we need more tests for the new methods here?

Reply:
It's not really much new here, but still can't hurt.
I'll add some.

Reply:
Done